File: 437. Path Sum III/solution.memory.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def pathSum(self, root: TreeNode, total: int) -> int:
        paths = 0
        if not root:
            return paths

        s = [(root, [])]
        while s:
            node, vals = s.pop()
            vals = [val + node.val for val in vals] + [node.val]

            if total in vals:
                paths += vals.count(total)
            if node.right:
                s.append((node.right, vals))
            if node.left:
                s.append((node.left, vals))

        return paths

File: 437. Path Sum III/solution.memory.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func sumStartAt(root *TreeNode, sum int, count *int) {
	if root == nil {
		return
	}
	if sum == root.Val {
		*count += 1
	}
	sumStartAt(root.Left, sum-root.Val, count)
	sumStartAt(root.Right, sum-root.Val, count)
}

func traverseAllNode(root *TreeNode, sum int, count *int) {
	if root == nil {
		return
	}

	sumStartAt(root, sum, count)
	traverseAllNode(root.Left, sum, count)
	traverseAllNode(root.Right, sum, count)
}

func pathSum(root *TreeNode, sum int) int {
	count := 0
	traverseAllNode(root, sum, &count)
	return count
}